src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cassert>
+#include <set>
+#include <sstream>
+
+#include "tracer.hpp"
+
+using namespace std;
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    thread_local std::shared_ptr<CTracer> tracer = nullptr;
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration);
+        return tracer;
+    }
+    std::shared_ptr<CTracer> CreateTracer(int preferredLayerDuration, const std::unordered_map<OpId, std::string>& opNames)
+    {
+        tracer = std::make_shared<CTracer>(preferredLayerDuration, opNames);
+        return tracer;
+    }

```

Comment:
Optional:
These 2 factory funcs (and the 2 constructors of CTracer) can be merged together. If you like, then I can tell you how (but the code will become a bit more tricky).

Reply:
The factories currently are only relevant when running QIR, but not if someone wants to consume/test the tracer directly from C++ (thus the constructors aren't private). I don't know whether running directly from C++ is a scenario we'd want to support in the future, though. For example, we could create the tracer on the stack for unit tests, but I'm not doing it... Don't have a strong opinion on this one, maybe, revisit later?